lib/nrf_modem_lib/nrf_modem_lib_trace.c

```diff
@@ -4,163 +4,218 @@
 }
 
-int nrf_modem_lib_trace_init(struct k_heap *trace_heap)
+void trace_thread_handler(void)
 {
 	int err;
-	t_heap = trace_heap;
-	trace_running = true;
 
-	err = trace_backend_init();
-	if (err) {
-		backend_initialized = false;
-		return -EBUSY;
-	}
+trace_reset:
+	k_sem_take(&trace_sem, K_FOREVER);
+
+	while (true) {
+		struct nrf_modem_trace_data *frags = NULL;
+		size_t n_frags = 0;
+
+		err = nrf_modem_trace_get(&frags, &n_frags);
+
+		/* Processing callback not called enough times (-EINPROGRESS),
+			* coredump (-ENODATA), or modem shutdown (-ESHUTDOWN).
+			* Unrecoverable errors.
+			*/
+		switch (err) {
+		case 0:
+			break;
+		case -EINPROGRESS:
+			/* Processing callback was not called enough times. */
+			LOG_ERR("nrf_modem_trace_get failed with err: %d", err);
+		case -ENODATA:
+			/* End of coredump. */
+		case -ESHUTDOWN:
+			/* Modem shutdown. */
+
+			err = trace_deinit();
+			if (err) {
+				LOG_ERR("trace_deinit failed with err: %d", err);
+			}
 
-	backend_initialized = true;
+			goto trace_reset;
+		default:
+			LOG_ERR("nrf_modem_trace_get failed with unexpected  err: %d", err);
 
-	return 0;
+			goto trace_reset;
+		}
+
+		for (size_t i = 0; i < n_frags; i++) {
+			err = trace_fragment_write(&frags[i]);
+			if (err) {
+				/* Trace backend unrecoverable failure so we deinit. */
+				err = trace_deinit();
+				if (err) {
+					LOG_ERR("trace_deinit failed with err: %d", err);
+				}
+
+				goto trace_reset;
+			}
+		}
+	}
 }
 
-int nrf_modem_lib_trace_start(enum nrf_modem_lib_trace_mode trace_mode)
+static int trace_init(void)
 {
-	if (!backend_initialized) {
-		return -EPERM;
-	}
+       int err;
 
-	if (nrf_modem_at_printf("AT%%XMODEMTRACE=1,%hu", trace_mode) != 0) {
-		return -EOPNOTSUPP;
-	}
+	   k_sem_take(&trace_done_sem, K_FOREVER);
 
-	trace_running = true;
+       err = trace_backend_init(nrf_modem_trace_processed);
+       if (err) {
+	       LOG_ERR("trace_backend_init failed with err: %d", err);
 
-	return 0;
-}
+	       return err;
+       }
 
-int nrf_modem_lib_trace_process(const uint8_t *data, uint32_t len)
-{
-	int err = 0;
+       err = nrf_modem_lib_trace_level_set(CONFIG_NRF_MODEM_LIB_TRACE_LEVEL_INIT);
+       if (err) {
+	       LOG_ERR("nrf_modem_lib_trace_mode_set failed with err: %d", err);
 
-	if (backend_initialized) {
-		struct trace_data_t *trace_data;
-		size_t bytes = sizeof(struct trace_data_t);
+	       return err;
+       }
 
-		trace_data = k_heap_alloc(t_heap, bytes, K_NO_WAIT);
+       k_sem_give(&trace_sem);
 
-		if (trace_data != NULL) {
-			trace_data->data = data;
-			trace_data->len = len;
+       return 0;
+}
 
-			k_fifo_put(&trace_fifo, trace_data);
-		} else {
-			LOG_ERR("trace_alloc failed");
+static void trace_init_callback(int init_err, void *ctx)
+{
+	int err;
 
-			err = -ENOMEM;
-		}
-	} else {
-		LOG_ERR("Modem trace received without initializing transport");
+	if (init_err) {
+		LOG_ERR("NRF_MODEM_LIB_ON_INIT failed with err: %d. Traces will not initialize.",
```

Comment:
just a suggestion. We already know from the log prefix we are in the trace lib. Also we don't care much here about the modem lib init failure, other that that it failed. 
```suggestion
		LOG_ERR("Initialization skipped due to modem lib failure",
```